File: c/linked-list.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    if (newNode == NULL) {
        printf("Node allocation failed.\n");
        exit(1);
    }
    newNode->data = data;
    newNode->next = NULL;

    return newNode;
}

struct Node* appendToTail(struct Node* head, int data) {
    struct Node* newNode = createNode(data);

    struct Node* current = head;
    if (current == NULL) {
        return newNode;
    }

    while(current->next != NULL) {
        current = current->next;
    }

    current->next = newNode;
    return head;
}

void printLL(struct Node* head) {
    struct Node* current = head;
    while (current != NULL) {
        printf("%d->", current->data);
        current = current->next;
    }
    printf("\n");
}

void freeLL(struct Node* head) {
    struct Node* current = head;
    struct Node* temp = NULL;

    while(current != NULL) {
        temp = current;
        current = current->next;
        printf("Free'd %d\n", temp->data);
        free(temp);
    }
}

int main(int argc, char** argv) {
    struct Node* head = NULL;

    if (head == NULL)
        printf("Head is NULL\n");
    else
        printf("Head is not NULL\n");


    for (int i = 0; i < 5; i++)
        head = appendToTail(head, i+1);    

    printLL(head);

    freeLL(head);

    printf("%d\n", head->data);

    if (head == NULL)
        printf("Head is NULL\n");
    else
        printf("Head is not NULL\n");

    return 0;
}

File: python/linked-list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def appendToTail(self, data):
        newNode = Node(data)
        
        current = self.head
        if current is None:
            self.head = newNode
            return

        while current.next:
            current = current.next

        current.next = newNode

    def appendToHead(self, data):
        newNode = Node(data)
        
        if self.head is None:
            self.head = newNode
            return
        
        newNode.next = self.head
        self.head = newNode

    def deleteFromTail(self):    
        node = self.head

        if node is None:
            # Empty LL
            return        
        elif node.next is None:
            self.head = None
            return
        
        while node.next.next:
            node = node.next

        node.next = None       

    def deleteFromHead(self):        
        if self.head is None:
            return
        
        self.head = self.head.next        

    def printLL(self):
        current = self.head
        ans = ""
        while current:
            ans += f"{current.data}->"
            current = current.next        
        print(ans)

    def size(self):
        current = self.head
        ans = 0
        while current:
            ans += 1
            current = current.next
        return ans

if __name__ == "__main__":
    # Initialize linked list
    myLL = LinkedList()

    for i in range(5):
        myLL.appendToTail(i)
    
    myLL.printLL()

    for i in range(10, 15):
        myLL.appendToHead(i)

    myLL.printLL()
    
    myLL.deleteFromTail()    
    myLL.printLL()
    myLL.deleteFromTail()
    myLL.printLL()
    myLL.deleteFromTail()    
    myLL.printLL()

    n = myLL.size()
    print(n)
    for _ in range(n):
        myLL.deleteFromHead()
        myLL.printLL()